crates/ty_server/tests/e2e/initialize.rs

```diff
@@ -474,3 +475,29 @@ fn register_multiple_capabilities() -> Result<()> {
 
     Ok(())
 }
+
+/// Tests that the server doesn't panic when `VIRTUAL_ENV` points to a non-existent directory.
+///
+/// See: <https://github.com/astral-sh/ty/issues/2031>
+#[test]
+fn missing_virtual_env_does_not_panic() -> Result<()> {
+    let workspace_root = SystemPath::new("project");
+
+    // This should not panic even though VIRTUAL_ENV points to a non-existent path
+    let mut server = TestServerBuilder::new()?
+        .with_workspace(workspace_root, None)?
+        .with_env_var("VIRTUAL_ENV", "/nonexistent/virtual/env/path")
+        .build()
+        .wait_until_workspaces_are_initialized();
+
+    let show_message_params = server.await_notification::<ShowMessage>();
+
+    insta::assert_json_snapshot!(show_message_params, @r#"
+    {
+      "type": 1,
+      "message": "Failed to load project rooted at <temp_dir>/project. Please refer to the logs for more details."
```

Comment:
Note this message is not "we crash and die" but rather "we are using the default fallback database that all python files outside the workspace have" (so hopefully graceful degradation, but enough of a degradation we probably want the user to look into it and fix it).